includes/API/Manifests/ApiLabkiHierarchyGet.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LabkiPackManager\API\Manifests;
+
+use ApiMain;
+use LabkiPackManager\Services\LabkiRepoRegistry;
+use LabkiPackManager\Services\ManifestStore;
+use Wikimedia\ParamValidator\ParamValidator;
+
+/**
+ * API endpoint to retrieve the computed hierarchy tree
+ * for a given repository and reference (branch/tag/commit).
+ *
+ * ## Action
+ * `labkiHierarchyGet`
+ *
+ * ## Purpose
+ * Returns a fully resolved, UI-ready hierarchy structure
+ * derived from the manifest.yml packs and dependencies.
+ * Does not include raw manifest or graph data.
+ *
+ * ## Example
+ * ```
+ * api.php?action=labkiHierarchyGet
+ *   &repo_url=https://github.com/Aharoni-Lab/labki-packs
+ *   &ref=main
+ *   &format=json
+ * ```
+ *
+ * ## Example Response
+ * ```json
+ * {
+ *   "repo_url": "https://github.com/Aharoni-Lab/labki-packs",
+ *   "ref": "main",
+ *   "hash": "19aba05e8751431c92890b569d4d2a5ef75a4194",
+ *   "hierarchy": { ... },
+ *   "meta": {
+ *     "schemaVersion": 1,
+ *     "timestamp": "20251027T213442",
+ *     "from_cache": true
+ *   }
+ * }
+ * ```
+ *
+ * @ingroup API
+ */
+final class ApiLabkiHierarchyGet extends ManifestApiBase {
+
+	private ?ManifestStore $manifestStore;
+
+	public function __construct(
+		ApiMain $main,
+		string $name,
+		?LabkiRepoRegistry $repoRegistry = null,
+		?ManifestStore $manifestStore = null
+	) {
+		parent::__construct($main, $name, $repoRegistry);
+		$this->manifestStore = $manifestStore;
+	}
+
+	/**
+	 * Execute the API request.
+	 */
+	public function execute(): void {
+		$params = $this->extractRequestParams();
+
+		// 1. Resolve and validate repository
+		$repoUrl = $this->resolveAndValidateRepo($params);
+		$repo = $this->repoRegistry->getRepo($repoUrl);
+
+		// 2. Resolve ref and refresh flag
+		$ref = $this->resolveRef($params, $repo);
+		$refresh = (bool)($params['refresh'] ?? false);
+
+		// 3. Get hierarchy and metadata from store (single call)
+		$store = $this->manifestStore ?? new ManifestStore($repoUrl, $ref);
+		$result = $store->getHierarchy($refresh);
+
+		if ($result === null) {
+			$this->dieWithError('labkipackmanager-error-fetch', 'fetch_error');
+		}
+
+		$meta = $result['meta'] ?? [];
+		$hierarchy = $result['hierarchy'] ?? [];
+
+		// 4. Output response
+		$out = $this->getResult();
+		$out->addValue(null, 'repo_url', $meta['repo_url'] ?? $repoUrl);
```

Comment:
the number of fallbacks in all these make me a bit nervous: why is it possible for this value to come from multiple places? if we expect a value to come from some place like the manifest store/hierarchy builder, why not throw an error there if we can't provide it? needing to propagate uncertainty through the whole codebase makes everything twice as verbose and error prone as it needs to be